### ISSUE TITLE
Provide a way for including an external config file for phar built

### ISSUE THREAD

**Issue by wowcut** [laravel-zero/laravel-zero#183]
It would be great to have a way to make the phar read an optional config file _from the filesystem, not from inside the phar._

But some things are special in phars, as all the file access maps to files _inside_ the phar. This means that everything inside `config/` will be included and read from _inside_ the phar archive - but it would be great to have a config file from _outside the phar_ to override the settings that are now "baked into" the phar archive. 

I found a solution for this, but it is not very clean as it works "outside" the Laravel framework with plain php, what is still ok, but I am sure there must be a better way, but atm I do not know at which point in time the framework loads files from `config/` - also I believe that the code that loads files from `config/` must be adopted to construct the path differently when in Phar execution mode. So this solution is just a hint into the direction how it might be implemented, certainly you can translate that to a proper patch.

I add this to the main command file:
```php
#!/usr/bin/env php
<?php

define('LARAVEL_START', microtime(true));

if (!empty($pharPath = \Phar::running(false))) {
    $file = dirname($pharPath) . '/config/'.basename(__FILE__).'.php';
    if ( file_exists($file) ) require $file;
}
```
Inside `config/name-of-command.php` I put only simple things, e.g.
```php
define('PATH_DATA', dirname(__DIR__) . '/data');
```
then for example in /config/config-for-laravel.php:
```php
defined('PATH_DATA') or define('PATH_DATA', dirname(__DIR__) . '/data');
return [ 'PATH_DATA' => PATH_DATA ];
```
So this can still be used with the Laravel way of accessing config variables, e.g `config()`.
We need to use `defined` because the "external" config file should be optional and if it not exists undefined constants will throw errors - but we still want to use them, when that file exists!

This works, but I am sure there might be some more elegant way to implement this is in a generic way for laravel-zero.

**Comment by nunomaduro** [comment]
Could you try to check the configuration file in `https://github.com/laravel-zero/laravel-zero/blob/stable/box.json`?

**Comment by wowcut** [comment]
This is the exact same configuration I have here already.

**Comment by nunomaduro** [comment]
@wowcut You should adapt it, in order to include external files.

**Comment by wowcut** [comment]
I do not see any way how changing that config can help, but maybe I am wrong. I see that one can exclude directories, but this would not help, as _it is ok_ to include `config/`.

**Comment by wowcut** [comment]
I do not want to include internal files. The point of this issue is including **external** files if they exist. SO one can deploy one phar but still use an external configuration file to adapt to diffrent environments. 
I guess, btw, that this would be also useful for `.env` files, but I did not test that.

**Comment by nunomaduro** [comment]
@wowcut A side note, not related to the issue: First of all, thanks for all your work so far on the feedback. Would be great if you could also write some content to improve our website: https://laravel-zero.com. https://github.com/laravel-zero/website.

**Comment by linuxjuggler** [comment]
Am not sure if my solution the cleanest one that you will see (as I am still working on refactoring it), but this is what I did, and I hope you will find it useful.

1- Installed filesystems and add the following config file

```php
return [
    'default' => 'local',
    'disks' => [
        'local' => [
            'driver' => 'local',
            'root' => sprintf('%s/.myAwesomeApp',getenv('HOME')),
        ],
    ],
];

```
as you can notice, am creating a new directory to hold my data within `getenv('HOME')`, there I'll be adding my generated `.env` file.

2- In `bootstrap/app.php` I added the following :

```php
/*
|--------------------------------------------------------------------------
| Set the correct path for the environment file
|--------------------------------------------------------------------------
|
| If the current directory has a .env file then we will use that instead
| of the global one, other wise the one under the user homepage.
|
 */

$app->instance('path.env', getEnvPath());

if (! file_exists(dirname(__DIR__) . DIRECTORY_SEPARATOR . '.env')) {
    $app->useEnvironmentPath(getEnvPath());
}
```

This will help me to load the local `.env` file am using within my code while coding since at that stage Laravel would not have registered the `environment` class so I can't check the environment to specify which file to load.

Am sure there is a better place for it, but this is just something I was working on today.

The content of the function `getEnvPath` is :

```php

if (! function_exists('getEnvPath')) {
    function getEnvPath()
    {
        return getenv('HOME') . DIRECTORY_SEPARATOR . '.myAwesomeApp';
    }
}
```
Which just return the path as string.

The key here is `useEnvironmentPath` which set the path where the `.env` file is located.
And the value which set by this function is used within the class `LoadEnvironmentVariables` class.

* Lastely, create a trait file which should be used within each and every commands that need the custom `.env` file, and put the following constructor

```php
    public function __construct()
    {
        parent::__construct();

        if(app()->environment('production') && ! Storage::disk('local')->exists('.env')) {
            // Do your logic here
        }

    }
```

This constructor will check for the environemnt and make sure to show a message indicate that they need to do something.

**Comment by linuxjuggler** [comment]
After small trial and error, I found out that the best place to check the existence of the `.env` file is within the `AppServiceProvider` or any service provider that you have.

but make sure to check it only when needed, otherwise, the validation will be executed all the time, and you will need to escape it sometime, like when you want to generate the `.env` file.

**Comment by ijpatricio** [comment]
Hello everyone.

Just made a project where the built phar may have a config file residing along, and merging values from this.

There's a configuration to enable/disable external file, and also for the filename.

It's initial stage, before eventually functionality getting to the core, just to get a picture of how it could be made.

https://github.com/ijpatricio/inspiring

Feedback appreciated.

Thanks!

**Comment by nunomaduro** [comment]
This issue has been fixed by @ijpatricio on Laravel Zero 5.8.
